plugins/lines/prism-lines.js

```diff
@@ -0,0 +1,266 @@
+							// Example: <span>foo\n</span>
+							return NO_SPLIT_NEWLINE;
+						}
+
+						// create a copy of the element without any child nodes (shallow copy)
+						var restElement = /** @type {Element} */ (element.cloneNode(false));
+
+						// append the rest of the split result of the current child to the new element
+						if (res.rest) {
+							restElement.append(res.rest);
+						}
+						// append all child nodes of the current child to the new element
+						transferLastChildNodes(element, restElement, l - i - 1);
+
+						return { newline: true, rest: restElement };
+					}
+				}
+
+				// This will be reached if the element has no children or none of the children contains a newline.
+				// Example: <span>foo</span>
+				return NO_SPLIT;
+
+			case Node.TEXT_NODE:
+				var text = /** @type {Text} */ (node);
+
+				var newlineIndex = text.data.indexOf('\n');
+				if (newlineIndex === -1) {
+					// This text does not contain a newline.
+					return NO_SPLIT;
+
+				} else if (newlineIndex === text.data.length - 1) {
+					// This text ends a line with nothing after it.
+					return NO_SPLIT_NEWLINE;
+
+				} else {
+					// This text contains a newlines and there are some characters after the first line.
+					var restText = text.splitText(newlineIndex + 1);
+					return { newline: true, rest: restText };
+				}
+
+			default:
+				// In all other cases, we will just assume that the node doesn't contribute to the textContent and
+				// therefore cannot contain newlines. Example: comments.
+				return NO_SPLIT;
+		}
+	}
+
+
+	/**
+	 * @typedef RegisterHookEnv
+	 * @property {HTMLPreElement} pre
+	 * @property {Prism.hooks.HookEnvironmentMap["before-insert"]} env
+	 * @property {LinesListener[]} listeners
+	 *
+	 * @typedef {{ [K in keyof Envs]?: (env: Envs[K]) => void }} LinesListener
+	 *
+	 * @typedef Envs
+	 * @property {BasicLinesEnv} onBefore
+	 * @property {BasicLinesEnv & NewLineEnv} onNewLine
+	 * @property {BasicLinesEnv & CompleteEnv} onComplete
+	 *
+	 * @typedef BasicLinesEnv
+	 * @property {HTMLPreElement} pre
+	 * @property {Prism.hooks.HookEnvironmentMap["before-insert"]} env
+	 *
+	 * @typedef NewLineEnv
+	 * @property {Element} line
+	 * @property {Node[]} childNodes
+	 *
+	 * @typedef CompleteEnv
+	 * @property {Node & ParentNode} lines
+	 */
+
+	var parser = new DOMParser();
+
+	Prism.hooks.add('before-insert', function (env) {
+		// works only for <code> wrapped inside <pre> (not inline)
+
+		var pre = /** @type {HTMLPreElement} */ (env.element.parentNode);
+		if (!pre || !/^pre$/i.test(pre.nodeName)) {
+			return;
+		}
+
+		// prepare the registration hook
+		/** @type {LinesListener[]} */
+		var listeners = [];
+		Prism.hooks.run('lines-register', /** @type {RegisterHookEnv} */({
+			pre: pre,
+			env: env,
+			listeners: listeners
+		}));
+
+		if (!listeners.length) {
+			// no plugin needs line wrappers, so we don't add any
+			return;
+		}
```

Comment:
Is it possible for CSS-only solutions to exist such that there could be no JS listeners but still worth wrapping every line?

Reply:
I added this is so the plugin does nothing by default. Without this, you just can't turn it off which is a problem because it is incompatible with Keep markup if kept nodes span multiple lines.